File: src/2025-02-23-deepclaude-try/javascript/exercises/practice/scale-generator/scale-generator.js
export class Scale {
  constructor(tonic) {
    this.tonic = this.normalizeTonic(tonic);
    this.baseScaleType = this.determineBaseScale();
    this.baseScale = this.baseScaleType === 'sharps' 
      ? ['A', 'A#', 'B', 'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#']
      : ['A', 'Bb', 'B', 'C', 'Db', 'D', 'Eb', 'E', 'F', 'Gb', 'G', 'Ab'];
    this.chromaticScale = this.buildChromaticScale();
  }

  normalizeTonic(tonic) {
    const normalized = tonic[0].toUpperCase() + tonic.slice(1).toLowerCase();
    return normalized.replace(/b/g, 'b');
  }

  determineBaseScale() {
    if (['C', 'A'].includes(this.tonic)) {
      return 'sharps';
    }

    const sharpMajors = ['G', 'D', 'A', 'E', 'B', 'F#'];
    const sharpMinors = ['E', 'B', 'F#', 'C#', 'G#', 'D#'];
    const flatMajors = ['F', 'Bb', 'Eb', 'Ab', 'Db', 'Gb'];
    const flatMinors = ['D', 'G', 'C', 'F', 'Bb', 'Eb'];

    if (flatMinors.includes(this.tonic) || flatMajors.includes(this.tonic)) {
      return 'flats';
    } else if (sharpMajors.includes(this.tonic) || sharpMinors.includes(this.tonic)) {
      return 'sharps';
    } else {
      throw new Error(`Invalid tonic: ${this.tonic}`);
    }
  }

  buildChromaticScale() {
    const index = this.baseScale.indexOf(this.tonic);
    if (index === -1) {
      throw new Error(`Tonic ${this.tonic} not found in ${this.baseScaleType} scale`);
    }
    return [...this.baseScale.slice(index), ...this.baseScale.slice(0, index)];
  }

  chromatic() {
    return this.chromaticScale;
  }

  interval(intervals) {
    let currentIndex = 0;
    const scale = [this.chromaticScale[0]];
    
    for (const interval of intervals) {
      const step = this.getStep(interval);
      currentIndex = (currentIndex + step) % 12;
      scale.push(this.chromaticScale[currentIndex]);
    }
    
    return scale;
  }

  getStep(interval) {
    switch(interval) {
      case 'm': return 1;
      case 'M': return 2;
      case 'A': return 3;
      default: throw new Error(`Invalid interval: ${interval}`);
    }
  }
}

File: src/2025-02-27-deepclaude37-rust/go/exercises/practice/scale-generator/scale_generator.go
package scale

import (
	"strings"
	"unicode"
)

var (
	sharpScale = []string{"A", "A#", "B", "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#"}
	flatScale  = []string{"A", "Bb", "B", "C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab"}

	sharpMajors = map[string]bool{"G": true, "D": true, "A": true, "E": true, "B": true, "F#": true}
	sharpMinors = map[string]bool{"E": true, "B": true, "F#": true, "C#": true, "G#": true, "D#": true}
	flatMajors  = map[string]bool{"F": true, "Bb": true, "Eb": true, "Ab": true, "Db": true, "Gb": true}
	flatMinors  = map[string]bool{"D": true, "G": true, "C": true, "F": true, "Bb": true, "Eb": true}
)

func normalizeTonic(tonic string) string {
	if len(tonic) == 0 {
		return ""
	}
	runes := []rune(tonic)
	first := string(unicode.ToUpper(runes[0]))
	rest := ""
	if len(runes) > 1 {
		rest = strings.ToLower(string(runes[1:]))
	}
	return first + rest
}

func determineScaleType(tonic string) []string {
	// Handle natural tonics first
	if tonic == "C" || tonic == "A" {
		return sharpScale
	}
	// Check flat keys before sharp keys
	if flatMajors[tonic] || flatMinors[tonic] {
		return flatScale
	}
	if sharpMajors[tonic] || sharpMinors[tonic] {
		return sharpScale
	}
	// Default to sharps
	return sharpScale
}

func indexOf(note string, scale []string) int {
	for i, n := range scale {
		if n == note {
			return i
		}
	}
	return -1
}

func rotateScale(base []string, index int) []string {
	return append(base[index:], base[:index]...)
}

func getSteps(interval string) []int {
	steps := make([]int, len(interval))
	for i, c := range interval {
		switch c {
		case 'm':
			steps[i] = 1
		case 'M':
			steps[i] = 2
		case 'A':
			steps[i] = 3
		}
	}
	return steps
}

func Scale(tonic, interval string) []string {
	// Normalize the tonic
	normalizedTonic := normalizeTonic(tonic)

	// Determine the base scale (sharp or flat)
	baseScale := determineScaleType(normalizedTonic)

	// Find the index of the tonic in the base scale
	index := indexOf(normalizedTonic, baseScale)
	if index == -1 {
		// Tonic not found in base scale
		return nil
	}

	// Rotate the base scale to start with the tonic
	rotated := rotateScale(baseScale, index)

	// If interval is empty, return the chromatic scale
	if interval == "" {
		return rotated
	}

	// Generate diatonic scale based on the interval pattern
	steps := getSteps(interval)
	currentIndex := 0
	scale := make([]string, 0, len(steps)+1)
	scale = append(scale, rotated[0])

	for _, step := range steps {
		currentIndex = (currentIndex + step) % 12
		scale = append(scale, rotated[currentIndex])
	}

	return scale
}
